contracts/zero-ex/contracts/src/features/ERC1155OrdersFeature.sol

```diff
@@ -0,0 +1,1003 @@
+        // `erc1155TokenId` matches the one specified in the order.
+        if (order.erc1155TokenProperties.length == 0) {
+            if (erc1155TokenId != order.erc1155TokenId) {
+                LibERC1155OrdersRichErrors.ERC1155TokenIdMismatchError(
+                    erc1155TokenId,
+                    order.erc1155TokenId
+                ).rrevert();
+            }
+        }
+
+        // Validate each property
+        for (uint256 i = 0; i < order.erc1155TokenProperties.length; i++) {
+            LibERC1155Order.Property memory property = order.erc1155TokenProperties[i];
+            // `address(0)` is interpreted as a no-op. Any token ID
+            // will satisfy a property with `propertyValidator == address(0)`.
+            if (address(property.propertyValidator) == address(0)) {
+                continue;
+            }
+
+            // Call the property validator and throw a descriptive error
+            // if the call reverts.
+            try property.propertyValidator.validateProperty(
+                address(order.erc1155Token),
+                erc1155TokenId,
+                property.propertyData
+            ) {} catch (bytes memory errorData) {
+                LibERC1155OrdersRichErrors.PropertyValidationFailedError(
+                    address(property.propertyValidator),
+                    address(order.erc1155Token),
+                    erc1155TokenId,
+                    property.propertyData,
+                    errorData
+                ).rrevert();
+            }
+        }
+    }
+
+    /// @dev Get the order info for an ERC1155 order.
+    /// @param order The ERC1155 order.
+    /// @return orderInfo Infor about the order.
+    function getERC1155OrderInfo(LibERC1155Order.ERC1155Order memory order)
+        public
+        override
+        view
+        returns (LibERC1155Order.OrderInfo memory orderInfo)
+    {
+        orderInfo.orderHash = getERC1155OrderHash(order);
+
+        {
+            LibERC1155OrdersStorage.Storage storage stor =
+                LibERC1155OrdersStorage.getStorage();
+            uint256 orderState = stor.orderState[orderInfo.orderHash];
+            orderInfo.remainingERC1155FillableAmount = order.erc1155TokenAmount
+                .safeSub128(uint128(orderState));
+
+            if (orderInfo.remainingERC1155FillableAmount == 0) {
+                orderInfo.status = LibERC1155Order.OrderStatus.FILLED;
+                return orderInfo;
+            }
+            if (orderState & CANCEL_BIT != 0) {
+                orderInfo.status = LibERC1155Order.OrderStatus.CANCELLED;
+                return orderInfo;
+            }
+        }
+
+        // Only buy orders with `erc1155TokenId` == 0 can be property
+        // orders.
+        if (order.erc1155TokenProperties.length > 0 &&
+                (order.direction != LibERC1155Order.TradeDirection.BUY_1155 ||
+                 order.erc1155TokenId != 0))
```

Comment:
oh yeah that's interesting

Reply:
Oh I guess the problem then is if `tokenId` is 0, it's ambiguous whether the order should only accept `tokenId = 0` or any `tokenId` that satisfies the properties

Reply:
Maybe we should use -1 for the wildcard? Or even a completely random number?

Reply:
Because those would be less likely to be a valid tokenId? 🤷

Reply:
`keccak('BAD_NFT_ID)`

Reply:
I could imagine someone creating a malicious 721 contract that uses whatever wildcard we choose, then uses our contracts to scam people

Reply:
> What if we allowed property checks on sell orders and buy orders with a token ID? It could be a way for a seller and buyer to guarantee that a token doesn't get devalued right before the fill.

Anyway, this functionality is achievable by having a property that checks for a specific tokenId + whatever additional validation

Reply:
> I could imagine someone creating a malicious 721 contract that uses whatever wildcard we choose, then uses our contracts to scam people

Like they can't just as easily mess with transfers to scam people on every settlement protocol.

Reply:
> > What if we allowed property checks on sell orders and buy orders with a token ID? It could be a way for a seller and buyer to guarantee that a token doesn't get devalued right before the fill.
> 
> Anyway, this functionality is achievable by having a property that checks for a specific tokenId + whatever additional validation


How if you're buying a sell order?

Reply:
> > > What if we allowed property checks on sell orders and buy orders with a token ID? It could be a way for a seller and buyer to guarantee that a token doesn't get devalued right before the fill.
> > 
> > 
> > Anyway, this functionality is achievable by having a property that checks for a specific tokenId + whatever additional validation
> 
> How if you're buying a sell order?

well we'd need to enable properties for sell orders, but that's separate from allowing tokenId + properties